### PR TITLE
:seedling:  Add deleting MachineDeploymentTopology to ClusterClass changes test

### PR DESCRIPTION
**Pull request by killianmuldoon**
Signed-off-by: killianmuldoon <kmuldoon@vmware.com>

Add a delete MachineDeploymentTopology test to the ClusterClass changes end to end test. This test should prevent regressions in this bug: https://github.com/kubernetes-sigs/cluster-api/pull/7403

#7404

### CODE REVIEW

File: test/e2e/clusterclass_changes.go [hunk 1]
```diff
@@ -177,7 +177,14 @@ func ClusterClassChangesSpec(ctx context.Context, inputGetter func() ClusterClas
 			Cluster:                   clusterResources.Cluster,
 			WaitForMachineDeployments: input.E2EConfig.GetIntervals(specName, "wait-worker-nodes"),
 		})
+		By("PASSED!")
 
+		By("Deleting a MachineDeploymentTopology in the Cluster Topology and wait for associated MachineDeployment to be deleted")
```
Comment (killianmuldoon) [inline test/e2e/clusterclass_changes.go:182]: Tested locally - passing on main and failing on main before the fix, so it looks like it should correctly catch a similar bug in future.

File: test/e2e/clusterclass_changes.go [hunk 2]
```diff
@@ -177,7 +177,14 @@ func ClusterClassChangesSpec(ctx context.Context, inputGetter func() ClusterClas
 			Cluster:                   clusterResources.Cluster,
 			WaitForMachineDeployments: input.E2EConfig.GetIntervals(specName, "wait-worker-nodes"),
 		})
+		By("PASSED!")
```
Comment (sbueringer) [inline test/e2e/clusterclass_changes.go:180]:
```suggestion
```

File: test/e2e/clusterclass_changes.go [hunk 3]
```diff
@@ -459,3 +466,43 @@ func rebaseClusterClassAndWait(ctx context.Context, input rebaseClusterClassAndW
...
+	Expect(input.ClusterProxy).ToNot(BeNil(), "Invalid argument. input.ClusterProxy can't be nil when calling deleteMachineDeploymentTopologyAndWait")
+	Expect(input.Cluster).ToNot(BeNil(), "Invalid argument. input.Cluster can't be nil when calling deleteMachineDeploymentTopologyAndWait")
+	Expect(len(input.Cluster.Spec.Topology.Workers.MachineDeployments)).To(BeNumerically(">", 0),
+		"Invalid Cluster. deleteMachineDeploymentTopologyAndWait requires at least one MachineDeploymentTopology to be defined in the Cluster topology")
+
+	log.Logf("Removing MachineDeploymentTopology from the Cluster Topology.")
+	patchHelper, err := patch.NewHelper(input.Cluster, input.ClusterProxy.GetClient())
+	// Remove the first MachineDeploymentTopology under input.Cluster.Spec.Topology.Workers.MachineDeployments
+	mdTopologyToDelete := input.Cluster.Spec.Topology.Workers.MachineDeployments[0]
+	input.Cluster.Spec.Topology.Workers.MachineDeployments = append(input.Cluster.Spec.Topology.Workers.MachineDeployments[:0], input.Cluster.Spec.Topology.Workers.MachineDeployments[0+1:]...)
```
Comment (sbueringer) [inline test/e2e/clusterclass_changes.go:490]:
```suggestion
	input.Cluster.Spec.Topology.Workers.MachineDeployments = input.Cluster.Spec.Topology.Workers.MachineDeployments{1:]
```
would this work as well?

File: test/e2e/clusterclass_changes.go [hunk 4]
```diff
@@ -459,3 +466,43 @@ func rebaseClusterClassAndWait(ctx context.Context, input rebaseClusterClassAndW
...
+
+	Expect(err).ToNot(HaveOccurred())
```
Comment (sbueringer) [inline test/e2e/clusterclass_changes.go:492]: let's move this up to l.488 + enter an empty line afterwards please

**Review by sbueringer — COMMENTED** [review]
Looks good, thx for the quick implementation

Just a few nits

**Comment by sbueringer** [comment]
Perfect, thx!

/test pull-cluster-api-e2e-full-main 
(this test is only run as part of full)

/lgtm

**Comment by sbueringer** [comment]
/assign @fabriziopandini 
PTAL :)

**Pull request merged**